text/0033-fake-account-packet/0033-fake-account-packet.md

```diff
@@ -0,0 +1,206 @@
+
+In order to maintain separation of concerns and to increase extensibility, this can been provided
+via a new trait; [`Fake`]
+(https://gitlab.com/Fraser999/Fake-Account-Packet/blob/master/src/fake.rs):
+
+```rust
+pub trait Fake {
+    fn fake(seed: &Seed) -> Self;
+
+    fn fake_version(seed: &Seed) -> u64;
+
+    fn to_rng_seed(seed: &Seed) -> Vec<u32> {
+        let mut new_seed = vec![];
+        let mut new_seed_elt = 0;
+        let mut count = 0;
+        for elt in &seed.0 {
+            new_seed_elt += (*elt as u32) << count;
+            count += 8;
+            if count == 32 {
+                new_seed.push(new_seed_elt);
+                count = 0;
+                new_seed_elt = 0;
+            }
+        }
+        if count != 32 {
+            new_seed.push(new_seed_elt);
+        }
+        new_seed
+    }
+}
+```
+
+This trait allows data types to implement their fake constructors as they see fit.  It also provides
+a helper function to allow easy conversion from a `sodiumoxide::crypto::sign::Seed` to a `Vec<u32>`
+which can be used by Rust random number generators.
+
+An [`ExampleAccount`]
+(https://gitlab.com/Fraser999/Fake-Account-Packet/blob/master/src/example_account.rs) which
+implements the `Fake` trait is shown here:
+
+```rust
+#[derive(Clone, PartialEq, Debug, RustcEncodable, RustcDecodable)]
+pub struct ExampleAccount {
+    pub id: u64,
+    pub other_stuff: Vec<u8>,
+}
+
+impl Fake for ExampleAccount {
+    fn fake(seed: &Seed) -> ExampleAccount {
+        let mut rng = IsaacRng::from_seed(&Self::to_rng_seed(seed));
+        let size: u8 = rng.gen();
+        ExampleAccount {
+            id: rng.gen(),
+            other_stuff: rng.gen_iter().take(size as usize).collect(),
+        }
+    }
+
+    fn fake_version(_seed: &Seed) -> u64 {
+        2
+    }
+}
+```
+
+and can be used in [`create_fake_account_packet()`]
+(https://gitlab.com/Fraser999/Fake-Account-Packet/blob/master/src/main.rs#L55-72):
+
+```rust
+fn create_fake_account_packet(id: XorName, session_ids: &[SessionId]) -> StructuredData {
+    let seed = get_seed(&id, &session_ids);
+    let (public_key, private_key) = sign::keypair_from_seed(&seed);
+    let type_tag = 0;
+    let fake_account = ExampleAccount::fake(&seed);
+    let data = serialisation::serialise(&fake_account).expect("Can't fail.");
+    let version = ExampleAccount::fake_version(&seed);
+    StructuredData::new(type_tag,
+                        id,
+                        version,
+                        data,
+                        vec![public_key],
+                        vec![],
+                        Some(&private_key))
+        .expect("Can't fail.")
+}
+```
+
+# Drawbacks
+
+This presents a very basic implementation.  It requires the Vaults to generate throwaway
+cryptographic keys in order to create a plausible fake - a relatively expensive operation.  For an
+attacker just looking to do a DOS attack, this might provide a vector.  Profiling may show that this
```

Comment:
Again, I don't think we're on the same page.  Even if we were to hash data names and/or session IDs, we'd still need the peers to create throwaway crypto keys I think?